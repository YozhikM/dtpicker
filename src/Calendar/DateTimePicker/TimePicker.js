/* @flow */

import React from 'react';
import SvgIcon from '../../SvgIcon/SvgIcon';
import TimeSelect from './TimeSelect';
import s from './Calendar.scss';
import type { Highlight } from '../helpers';

type Show = 'button' | 'timeSelect';

type Props = {|
  show?: Show,
  date?: Highlight,
  onSetTime?: (time: Highlight) => void
|};

type State = {|
  show: 'button' | 'timeSelect'
|};

class TimePicker extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      show: this.props.show || 'button'
    };
  }

  onClickClock = () => {
    this.setState({ show: 'timeSelect' });
  };

  onSubmitTime = (time: Highlight) => {
    const { onSetTime } = this.props;
    this.setState({ show: 'button' });
    if (onSetTime) onSetTime(time);
  };

  render() {
    const { date } = this.props;
    const { show } = this.state;

    if (show === 'timeSelect') {
      return <TimeSelect value={date} onSubmit={this.onSubmitTime} />;
    }

    return (
      <div className={s.time_picker}>
        <button onClick={this.onClickClock}>
          <SvgIcon file="clock" />
        </button>
      </div>
    );
  }
}

export default TimePicker;
