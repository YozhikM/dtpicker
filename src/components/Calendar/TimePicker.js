/* @flow */

import React from 'react';
import SvgIcon from '../SvgIcon';
import TimeSelect from '../DateTimePicker/TimeSelect';
import s from './Calendar.scss';

type Props = {
  onChangeDate?: (date: Date) => void,
  activeDates: Date,
  date: Date,
  onSubmit?: any => void
};

type State = {
  date: Date,
  show: 'button' | 'timePicker'
};

class TimePicker extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      date: this.props.date,
      show: 'button'
    };
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.date !== this.props.date) {
      this.setState({ date: nextProps.date });
    }
  }

  onChangeDate = () => {
    const { onChangeDate } = this.props;
    if (onChangeDate) onChangeDate(this.state.date);
  };

  onClickClock = () => {
    this.setState({ show: 'timePicker' });
  };

  onSubmitTime = () => {
    this.setState({ show: 'button' })
  };

  render() {
    const { show, date } = this.state;
    if (show === 'timePicker') {
      return <TimeSelect {...this.props} value={date} onSubmit={this.onSubmitTime} onChange={this.onChangeDate} />;
    }
    return (
      <div className={s.time_picker}>
        <button onClick={this.onClickClock}>
          <SvgIcon file="clock" />
        </button>
      </div>
    );
  }
}

export default TimePicker;
