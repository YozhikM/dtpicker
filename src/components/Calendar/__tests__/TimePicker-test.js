/* @flow */

import React from 'react';
import TimePicker from '../TimePicker';

function findBtn(wrapper) {
  return wrapper.find('button');
}

function findTimeSelect(wrapper) {
  return wrapper.find('TimeSelect');
}

describe('TimePicker', () => {
  describe('render()', () => {
    it('basic', () => {
      expect(shallow(<TimePicker date={new Date(2010, 10, 10)} />)).toMatchSnapshot();
    });

    it('show TimeSelect', () => {
      expect(
        shallow(<TimePicker date={new Date(2010, 10, 10)} show={'timeSelect'} />)
      ).toMatchSnapshot();
    });
  });

  it('componentWillReceiveProps()', () => {
    const wrapper = shallow(<TimePicker />);
    wrapper.setProps({ date: new Date(2001, 8, 4, 10, 13, 12) });
    expect(wrapper.state()).toEqual({ date: new Date('2001-09-04T03:13:12.000Z'), show: 'button' });
  });

  it('onChangeDate()', () => {
    const date = new Date(2000, 1, 1, 0, 0, 0);
    const spy = jest.fn();
    const wrapper = shallow(<TimePicker date={date} onChangeDate={spy} show={'timeSelect'} />);
    wrapper
      .find('TimeSelect')
      .props()
      .onChange(date);
    expect(spy).toBeCalled();
    const callbackValue = spy.mock.calls[0][0];
    expect(callbackValue).toEqual(date);
  });

  describe('user interactions', () => {
    it('onClickClock()', () => {
      const wrapper = shallow(<TimePicker />);
      findBtn(wrapper).simulate('click');
      expect(wrapper.state('show')).toBe('timeSelect');
    });

    it('onSubmitTime()', () => {
      const wrapper = shallow(<TimePicker />);
      wrapper.setState({ show: 'timeSelect' });
      findTimeSelect(wrapper).simulate('submit');
      expect(wrapper.state('show')).toBe('button');
    });
  });
});
