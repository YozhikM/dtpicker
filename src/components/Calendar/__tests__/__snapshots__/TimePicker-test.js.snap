// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimePicker render() basic 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    <SvgIcon
      file="clock"
    />
  </button>
</div>
`;

exports[`TimePicker render() show TimeSelect 1`] = `
<TimeSelect
  onChange={[Function]}
  onSubmit={[Function]}
  value={2010-11-09T18:00:00.000Z}
/>
`;
